### PR TITLE
Feature/reporting framework

### PR DESCRIPTION
**Pull request by sauroter**

### CODE REVIEW

File: src/main/java/org/varroa/scenario/ioc/ActionMappingProvider.java [hunk 1]
```diff
@@ -17,19 +17,24 @@
...
+            actionMapping = new ActionMapping(commands, actionCount);
+        }
+        return actionMapping;
     }
 
-    public void setCommands(final @NotNull ImmutableList<Command> commands, final int actionCount) {
+    public synchronized void setCommands(final @NotNull ImmutableList<Command> commands, final int actionCount) {
         this.commands = commands;
         this.actionCount = actionCount;
+        this.actionMapping = null;
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/scenario/ioc/ActionMappingProvider.java]: Creating the actionMapping here would be cleaner

File: src/main/java/org/varroa/scenario/ioc/ActionMappingProvider.java [hunk 2]
```diff
@@ -17,19 +17,24 @@
 
     private @Nullable ImmutableList<Command> commands = null;
     private int actionCount;
+    private ActionMapping actionMapping;
 
     @Override
-    public ActionMapping get() {
+    public synchronized ActionMapping get() {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/scenario/ioc/ActionMappingProvider.java]: Is synchronized necessary?
I think a volatile field actionMapping would be enough

Comment (sauroter) [reply src/main/java/org/varroa/scenario/ioc/ActionMappingProvider.java]: you are right

File: src/main/java/org/varroa/scenario/ioc/ActionMappingProvider.java (hunk 1 above)
Comment (sauroter) [reply src/main/java/org/varroa/scenario/ioc/ActionMappingProvider.java]: will do

File: src/main/java/org/varroa/scenario/LifeCycle.java [hunk 3]
```diff
@@ -69,4 +72,21 @@ public int getActionCount() {
...
+        return containsCommand(this.commands, command);
+    }
+
+    private static boolean containsCommand(final @NotNull List<Command> commands, final @NotNull Command command) {
+        for (final Command c : commands) {
+            if (command.equals(c)) {
+                return true;
+            }
+            if (c instanceof ForCommand) {
+                return containsCommand(((ForCommand) c).getCommands(), command);
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/scenario/LifeCycle.java]: imho this does not work, this should only return if `true`

File: src/main/java/org/varroa/reporting/Reporter.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                                    command, new ActionReportWithId(actionId, actionReport.getValue()));
+                        })
+                        .collect(groupingBy(
+                                Map.Entry::getKey,
+                                mapping(Map.Entry::getValue, toList())));
+
+        final List<ActionReportWithCommand> aggregatedReports =
+                groupByCommand.entrySet().stream()
+                        .map(groupedReports ->
+                                {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/reporting/Reporter.java]: minor: fformatting

File: src/main/java/org/varroa/scenario/LifeCycle.java (hunk 3 above)
Comment (sauroter) [reply src/main/java/org/varroa/scenario/LifeCycle.java]: done

File: src/main/java/org/varroa/reporting/Reporter.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                                    final ActionReportWithId mergedReports =
+                                            ActionReportWithId.merge(groupedReports.getValue());
+
+                                    return new AbstractMap.SimpleEntry<>(
+                                            mergedReports.id,
+                                            new ActionReportWithCommand(
+                                                    groupedReports.getKey(), mergedReports.getActionReport()));
+                                }
+                        )
+                        .sorted(Comparator.comparingInt((Map.Entry::getKey)))
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/reporting/Reporter.java]:
```suggestion
                        .sorted(Comparator.comparingInt(Map.Entry::getKey))
```

File: src/main/java/org/varroa/reporting/Reporter.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+        final List<ActionReportWithCommand> mappedReports = actionReports.entrySet().stream()
+                .map(actionReport -> {
+                    final Integer actionId = actionReport.getKey();
+                    final Command command = actionMapping.getCommand(actionId);
+
+                    return new AbstractMap.SimpleEntry<>(
+                            actionId, new ActionReportWithCommand(command, actionReport.getValue()));
+                })
+                .sorted(Comparator.comparingInt((Map.Entry::getKey)))
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/reporting/Reporter.java]:
```suggestion
                .sorted(Comparator.comparingInt(Map.Entry::getKey))
```

**Pull request merged**